### PR TITLE
Improve support for high-resolution stats

### PR DESCRIPTION
**Pull request by andygrundman**
NOTE: Users of this library may need to make changes. I've also submitted a sister patch for moonlight-qt that supports these changes and implements a few new features like an audio overlay.

* This patch adds a new microsecond-resolution function call, LiGetMicroseconds(), to complement the existing LiGetMillis(). Many variables used by stats have been updated to work at this higher resolution and now provide better results when displaying e.g. sub-millisecond frametime stats. To try and avoid confusion, variables that now contain microseconds have been renamed with a suffix of 'Us', and those ending in 'Ms' contain milliseconds. I originally experimented with nanoseconds but it felt like overkill for our needs.

* Since this library is designed to be mostly standalone, I reorganized Platform.c a bit to make it compatible with SDL's GetTicks64(), which starts its ticker at program start. A lot of the stats here are used with those in moonlight-qt so I tried to simplify the functions as much as possible. Each platform now has its own few smaller functions, instead of trying to fit a complex set of ifdef's inside the same function.

* I added a simple gtest suite for the Platform.c changes, and this test suite should be easy to extend to other areas of the code.

Internal API:
void Plt_TicksInit(void);          // store initial timestamp
uint64_t Plt_GetTicks64_us(void);  // The most precision, in microseconds
uint64_t Plt_GetTicks64_ms(void);  // Plt_GetTicks64_ms returns the value in milliseconds
uint64_t Plt_GetTicks64(void);     // Alias to _ms() and compatible with SDL_GetTicks64 (ms since program start)
uint64_t PltGetMillis(void);       // replaced by Plt_GetTicks64_ms()

Public API in Limelight.h:
uint64_t LiGetMicroseconds(void);
uint64_t LiGetMillis(void);
PRTP_AUDIO_STATS LiGetRTPAudioStats(void);	// provides access to RTP data for the overlay stats PRTP_VIDEO_STATS LiGetRTPVideoStats(void);

### CODE REVIEW

**Comment by andygrundman** [comment]
I think the build issues might be fixed now for Win/Mac/Linux.

I wanted to mention  a new error I ran into thanks to the `-fanalyzer` mode on one of the Linux builds. It ended up really impressing me with what the compiler is able to spot now.  In [this bit of code](https://github.com/moonlight-stream/moonlight-common-c/blob/8795c6c28d61671cedbf473e478700ae3e1f491a/src/RtspConnection.c#L667-L696) there is a loop from 0-7 accessing an array of speaker channels.

``` 
for (i = 0; i < opusConfig->channelCount; i++) {
        opusConfig->mapping[i] = CHAR_TO_INT(*paramStr);
        paramStr++;
}
```

The code was correct of course, but the compiler had no way to know what `channelCount` might be and that we wouldn't overflow the array. This caused the following error that was pretty confusing, especially when I hadn't changed anything nearby. (I think it was just part of an appveyor gcc update since the last time this got built.) Luckily `-Werror=stringop-overflow=` was a helpful starting point.
```
/home/appveyor/projects/moonlight-common-c/src/RtspConnection.c: In function ‘parseOpusConfigFromParamString’:
/home/appveyor/projects/moonlight-common-c/src/RtspConnection.c:691:32: error: writing 1 byte into a region of size 0 [-Werror=stringop-overflow=]
  691 |         opusConfig->mapping[i] = CHAR_TO_INT(*paramStr);
      |                                ^
In file included from /home/appveyor/projects/moonlight-common-c/src/Platform.h:81,
                 from /home/appveyor/projects/moonlight-common-c/src/Limelight-internal.h:3,
                 from /home/appveyor/projects/moonlight-common-c/src/RtspConnection.c:1:
/home/appveyor/projects/moonlight-common-c/src/Limelight.h:334:19: note: at offset 8 into destination object ‘mapping’ of size 8
  334 |     unsigned char mapping[AUDIO_CONFIGURATION_MAX_CHANNEL_COUNT];
      |                   ^~~~~~~
```

The fix turned out to be exactly the kind of safety code that should exist here. Now it's much more obvious how the loop variable works and the compiler was happy with it. This new if statement is at the beginning of the function, about 20 lines away from the loop. Pretty neat!
```
    if (channelCount > AUDIO_CONFIGURATION_MAX_CHANNEL_COUNT) {
        channelCount = AUDIO_CONFIGURATION_MAX_CHANNEL_COUNT;
    }
    opusConfig->channelCount = channelCount;
```

**Comment by Rhea4life** [comment]
Hey I will not be using this email for you anymore so please just email me at
***@***.***

please and thank you.

________________________________

**Comment by andygrundman** [comment]
I've removed the test suite in my patch for a few reasons:
* couldn't get it to build on win64, some MSVC/cmake/appveyor conflict, I dunno.
* Probably weird to have a C++ test suite for a C library. 
* the tests were time-based which means they will break and annoy someone in the future

File: src/Platform.c [hunk 1]
```diff
@@ -464,24 +464,162 @@ void PltWaitForConditionVariable(PLT_COND* cond, PLT_MUTEX* mutex) {
 #endif
 }
 
-uint64_t PltGetMillis(void) {
+//// Begin SDL time emulation
+
+// These functions mostly emulate SDL_GetTicks64(), because this library doesn't
+// use SDL, but timing data needs to be compatible with SDL-supplied data in moonlight-qt.
```
Comment (cgutman) [inline src/Platform.c:470]: I'm not really sure I understand the motivation of trying to match SDL. The epoch of `PltGetMillis()` has always been implementation-defined, and the app can normalize it in whatever way it chooses by calling `PltGetMillis()` .

We can't promise SDL-style times, because we have no control over when the client chooses to call `LiStartConnection()`. It's quite likely that the SDL timebase and our timebase will differ significantly, since successful network requests to the host need to happen before `LiStartConnection()` can be called to begin negotiating the connection.

File: src/Platform.c [hunk 2]
```diff
@@ -464,24 +464,162 @@ void PltWaitForConditionVariable(PLT_COND* cond, PLT_MUTEX* mutex) {
...
+void Plt_TicksInit(void) {
+    if (ticks_started) {
+        return;
+    }
+    ticks_started = true;
+    BOOL rc = QueryPerformanceFrequency(&ticks_per_second);
+    QueryPerformanceCounter(&start_ticks);
+}
+
+uint64_t Plt_GetTicks64_us(void) {
```
Comment (cgutman) [inline src/Platform.c:489]: I think I'd prefer to match the external API nomeclature here `PltGetMicroseconds()`, especially since I don't think we should promise the SDL timebase semantics and we don't use underscores in our API names elsewhere.

File: src/Limelight.h [hunk 3]
```diff
@@ -848,6 +851,36 @@ int LiGetPendingAudioFrames(void);
...
+    uint32_t packetCountAudio;         // total audio packets
+    uint32_t packetCountFec;           // total packets of type FEC
+    uint32_t packetCountFecRecovered;  // a packet was saved
+    uint32_t packetCountFecFailed;     // tried to recover but too much was lost
+    uint32_t packetCountOOS;           // out-of-sequence packets
+    uint32_t packetCountInvalid;       // corrupted packets, etc
+    uint32_t packetCountFecInvalid;    // invalid FEC packet
+} RTP_AUDIO_STATS, *PRTP_AUDIO_STATS;
+
+PRTP_AUDIO_STATS LiGetRTPAudioStats(void);
```
Comment (cgutman) [inline src/Limelight.h:866]: We should have these functions return const pointers to avoid the possibility of the data being modified.

File: src/RtspConnection.c [hunk 4]
```diff
@@ -665,6 +665,9 @@ static int parseOpusConfigFromParamString(char* paramStr, int channelCount, POPU
     int i;
 
     // Set channel count (included in the prefix, so not parsed below)
+    if (channelCount > AUDIO_CONFIGURATION_MAX_CHANNEL_COUNT) {
+        channelCount = AUDIO_CONFIGURATION_MAX_CHANNEL_COUNT;
+    }
```
Comment (cgutman) [inline src/RtspConnection.c:670]: Thanks for the fix. I took care of it slightly differently (failing rather than clamping to the max).

File: src/Limelight.h [hunk 5]
```diff
@@ -154,21 +154,20 @@ typedef struct _DECODE_UNIT {
...
-    // can be calculated by LiGetMillis() - enqueueTimeMs.
-    uint64_t enqueueTimeMs;
+    // can be calculated. This value is in microseconds.
+    uint64_t enqueueTimeUs;
 
     // Presentation time in milliseconds with the epoch at the first captured frame.
     // This can be used to aid frame pacing or to drop old frames that were queued too
     // long prior to display.
-    unsigned int presentationTimeMs;
+    uint64_t presentationTimeMs;
```
Comment (cgutman) [inline src/Limelight.h:170]: I generally try to avoid breaking API changes if possible, but if we're going to do them here (and it's probably fine), let's just go all the way and change `presentationTimeMs` -> `presentationTimeUs` too.

**Review by cgutman — CHANGES_REQUESTED** [review]

File: src/Platform.c (hunk 1 above)
Comment (andygrundman) [reply src/Platform.c:470]: I remember why I said this: it's because there are a lot of calls in -qt to `SDL_GetTicks()` for tracking time, as well as calls to `LiGetMillis()`. Since they both offer millisecond-resolution it doesn't really matter which is used as long as they aren't compared with each other for timing purposes. I don't think they are ever mixed, but the potential for confusion is there. I wanted to replace both calls with a microsecond version, so when it came to writing the Li version it made sense to me to just use the same "counter starts at first API call" plan that SDL uses. I'd suggest that we go with the version in my patch because it's well-tested code from SDL. And then we remove my reference to SDL as well as the start time of the program. How about this wording for the 2 functions in Limelight.h?
```
// This function returns a time in microseconds with an implementation-defined epoch. It should only ever be compared with the return value from a previous call to itself.
```

File: src/Platform.c (hunk 2 above)
Comment (andygrundman) [reply src/Platform.c:489]: Sounds good, I will put back `PltGetMillis()` and rename this to `PltGetMicroseconds()`.

File: src/Limelight.h (hunk 3 above)
Comment (andygrundman) [reply src/Limelight.h:866]: Agreed, will add.

File: src/Limelight.h (hunk 5 above)
Comment (andygrundman) [reply src/Limelight.h:170]: I probably screwed up here. I double-checked and this value seems to only be used internally during RTP processing and FEC recovery. I left this one at milliseconds because the original value comes from the server side at `newEntry->presentationTimeMs = packet->timestamp / PTS_DIVISOR;` and seemed self-contained to the RTP code. I think reverting to unsigned int would be fine.

Sort of related side note here, I have been wondering if we ought to be using the RTP timestamps as a way to sync up the video and audio streams, which I've seen done in other programs. I think we get away with it due to the minimal buffering, but it may still cause issues. I think this would have to start at giving the [audio](https://github.com/LizardByte/Sunshine/blob/master/src/stream.cpp#L1648) and [video](https://github.com/LizardByte/Sunshine/blob/master/src/stream.cpp#L1480-L1482) packets the same timestamp source/format. We can discuss this another time, though.

Comment (cgutman) [reply src/Limelight.h]: > Sort of related side note here, I have been wondering if we ought to be using the RTP timestamps as a way to sync up the video and audio streams, which I've seen done in other programs.

It's fairly common for most applications, but real-time game streaming has incredibly stringent latency requirements. Delaying video by 40ms or whatever to sync up with audio would be totally fine for something like Zoom, but would likely make Moonlight unplayable.

File: src/Platform.c [hunk 6]
```diff
@@ -464,24 +464,157 @@ void PltWaitForConditionVariable(PLT_COND* cond, PLT_MUTEX* mutex) {
...
+    has_monotonic_time = true;
+    start = mach_absolute_time();
+}
+
+uint64_t PltGetMicroseconds(void) {
+    if (!ticks_started) {
+        PltTicksInit();
+    }
+    const uint64_t now = mach_absolute_time();
+    return (((now - start) * mach_base_info.numer) / mach_base_info.denom) / 1000;
```
Comment (cgutman) [inline src/Platform.c:517]: `mach_absolute_time()` is apparently discouraged now due to device fingerprinting concerns. It looks like `clock_gettime_nsec_np()` is the newer function they're recommending. It also avoids us having to deal with `mach_timebase_info()` too.

File: src/Limelight.h (hunk 5 above)
Comment (cgutman) [reply src/Limelight.h]: > I probably screwed up here. I double-checked and this value seems to only be used internally during RTP processing and FEC recovery. I left this one at milliseconds because the original value comes from the server side at newEntry->presentationTimeMs = packet->timestamp / PTS_DIVISOR; and seemed self-contained to the RTP code. I think reverting to unsigned int would be fine.

Sorry I meant that we should convert this to microseconds and leave it as `uint64_t`. I know it currently comes out of RTP timestamps which limits the resolution of the value, but I think we should make it microseconds to be consistent with the other values and to leave room for future timestamp improvements without having to break the API again.

Comment (andygrundman) [reply src/Limelight.h]: > It's fairly common for most applications, but real-time game streaming has incredibly stringent latency requirements. Delaying video by 40ms or whatever to sync up with audio would be totally fine for something like Zoom, but would likely make Moonlight unplayable.

Oh yeah, video has to be the priority, so in our case I think you would want to check video/audio pts for drift and adjust the audio stream if it was ahead or behind the video. If you're ahead of the video, play silence until the video catches up, and if you're behind, skip some samples (or possibly use some time/pitch adjustment to play it faster. That's how the Bluetooth latency reduction works in macOS Game Mode for example. I will experiment more with this, since the pts values from Sunshine are different I am not sure if this is even an issue right now. I do know that audio DACs can drift by a surprising amount, having worked on audio sync for network music players  many years ago.

> Sorry I meant that we should convert this to microseconds and leave it as uint64_t. I know it currently comes out of RTP timestamps which limits the resolution of the value, but I think we should make it microseconds to be consistent with the other values and to leave room for future timestamp improvements without having to break the API again.

Roger, I am glad you said this, it's the right way to go. When backing it out I realized I had changed it because leaving it as unsigned int and doing some annoying casting just felt bad.

File: src/Platform.c (hunk 6 above)
Comment (andygrundman) [reply src/Platform.c:517]: Good catch, I had thought I looked at [these docs](https://developer.apple.com/documentation/kernel/1462446-mach_absolute_time) which has a big yellow warning on it warning of App Store privacy-related things if you use it.  I will switch to the other one.

Comment (andygrundman) [reply src/Platform.c:517]: Done, I also moved the `has_monotonic_time` variable into the final else block (Linux/Unix) since no other platform needs it. I didn't squash my commits this time but let me know if you would like the patch as 1 commit.

File: src/Limelight.h (hunk 5 above)
Comment (andygrundman) [reply src/Limelight.h:170]: Oops, it's always MSVC catching my bugs: I forgot to fix `queue->lastOosFramePresentationTimestamp` which was still 32-bit. I am not sure why this isn't a warning on other compilers... I really appreciate the appveyor builds. :)

**Comment by andygrundman** [comment]
Apologies for the delay, I am not entirely sure if marking everything resolved lets you know or if I am supposed to use "re-request review".
